rayon-core/src/join/mod.rs

```diff
@@ -20,6 +20,13 @@ mod test;
 /// only execute code in parallel when there are idle CPUs to handle
 /// it.
 ///
+/// When `join` is called from outside the thread pool, the calling
+/// thread will block while the closures execute in the pool.  When
+/// `join` is called within the pool, the calling thread still actively
+/// participates in the thread pool.  It will try to execute its own
+/// closures first, but if one is stolen then it will look for other
+/// work while waiting for that result to come back.
```

Comment:
Hmm.  I actually don't like that queue example much, because it will incur linear stack usage.  I don't think it is a good pattern for us to suggest with `join`.